### PR TITLE
Add V2Compatibility middleware

### PR DESCRIPTION
**Pull request by nambrot**
### Description

Adds V2Compatbility middleware that allows users to use the `V2CompatibilityRouter` in place of v2 mailboxes

### CODE REVIEW

File: solidity/contracts/Router.sol [hunk 1]
```diff
@@ -73,6 +73,21 @@ abstract contract Router is AbacusConnectionClient, IMessageRecipient {
         _enrollRemoteRouter(_domain, _router);
     }
 
+    /**
+     * @notice Batch version of `enrollRemoteRouter`
+     * @param _domains The domaisn of the remote Application Routers
```
Comment (tkporter) [inline solidity/contracts/Router.sol:78]: nit: `domaisn`

File: solidity/contracts/middleware/V2CompatibilityRouter.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+import {Router} from "../Router.sol";
+import {TypeCasts} from "../libs/TypeCasts.sol";
+import {IMessageRecipient} from "../../interfaces/IMessageRecipient.sol";
+
+/*
+ * @title The Hello World App
```
Comment (tkporter) [inline solidity/contracts/middleware/V2CompatibilityRouter.sol:9]: nit: seems stale?

File: solidity/contracts/middleware/V2CompatibilityRouter.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    function initialize(
+        address _owner,
+        address _abacusConnectionManager,
+        address _interchainGasPaymaster
+    ) public initializer {
+        // Transfer ownership of the contract to deployer
+        _transferOwnership(_owner);
+        // Set the addresses for the ACM and IGP
+        // Alternatively, this could be done later in an initialize method
```
Comment (tkporter) [inline solidity/contracts/middleware/V2CompatibilityRouter.sol:24]: ditto feels stale given that this is an initialize method

File: solidity/contracts/middleware/V2CompatibilityRouter.sol [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+     * @notice Adds a domain ID mapping from v1/v2 domain IDs and vice versa
+     * @param _v1Domains An array of v1 domain IDs
+     * @param _v2Domains An array of v2 domain IDs
+     */
+    function mapDomains(
+        uint32[] calldata _v1Domains,
+        uint32[] calldata _v2Domains
+    ) external onlyOwner {
+        for (uint256 i = 0; i < _v1Domains.length; i += 1) {
+            v1ToV2Domain[_v1Domains[i]] = _v2Domains[i];
```
Comment (tkporter) [inline solidity/contracts/middleware/V2CompatibilityRouter.sol:39]: think it's worth any input validation? e.g. making sure the mapping isn't already set or something

File: solidity/contracts/middleware/V2CompatibilityRouter.sol [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                    abi.encode(
+                        TypeCasts.addressToBytes32(msg.sender),
+                        _recipientAddress,
+                        _messageBody
+                    )
+                )
+            );
+    }
+
+    /**
```
Comment (tkporter) [inline solidity/contracts/middleware/V2CompatibilityRouter.sol:69]: You know I'm always a natspec advocate 😛 
But don't feel super strongly here given this is mostly a temporary thing

**Review by tkporter — APPROVED** [review]
Agree with just calling dispatch

File: solidity/contracts/middleware/V2CompatibilityRouter.sol (hunk 4 above)
Comment (nambrot) [reply solidity/contracts/middleware/V2CompatibilityRouter.sol:39]: Yeah I was thinking about it, but felt like i wanted to just maintain full flexibility for the owner

File: solidity/contracts/middleware/V2CompatibilityRouter.sol (hunk 5 above)
Comment (nambrot) [reply solidity/contracts/middleware/V2CompatibilityRouter.sol:69]: Man even for internal functions like this :P

**Pull request merged**